[TySan] Fix false positives with derived classes

gbMattN opened the pull request:
Fixes issue #125079 as well as another case I discovered while trying to build LLVM with TySan.
The case when the access has an offset needs some slightly different legwork to the average check.

gbMattN commented:
@fhahn nudge for review

gbMattN commented:
@fhahn Do you think it would be possible to get this into a release candidate before llvm20 releases?

fhahn commented on compiler-rt/lib/tysan/tysan.cpp:
```diff
@@ -102,20 +102,8 @@ static tysan_type_descriptor *getRootTD(tysan_type_descriptor *TD) {
-    TDA = TDA->Member.Base;
-  }
-
+bool walkAliasTree(tysan_type_descriptor *TDA, tysan_type_descriptor *TDB,
```
Should still be static?

```suggestion
static bool walkAliasTree(tysan_type_descriptor *TDA, tysan_type_descriptor *TDB,
```

fhahn replied:
still pending?

fhahn reviewed (commented):
> Fixes issue https://github.com/llvm/llvm-project/issues/125079 as well as another case I discovered while trying to build LLVM with TySan.
The case when the access has an offset needs some slightly different legwork to the average check.

It would be great if you could spell out the differences, and ideally also make this clear in the code comments.

gbMattN commented:
> It would be great if you could spell out the differences, and ideally also make this clear in the code comments.

I've added a comment into the code to explain the rational. I will also explain it here.
When you walk up the alias tree, the TDs go from structs to members. If you are accessing a TD in shadow with an offset, you have either an access violation, or are accessing a large struct TD. If you walk up the alias tree of the type you are accessing with, you won't find that struct, since you are accessing it with a TD for a member variable _inside_ said struct. 

Therefore we should instead walk up the alias tree of this large struct to find the type we are accessing, and then check to see if the offset of that type in the struct is equal to the offset of our read in shadow memory. Since this method of checking is the opposite to the default way of checking, I've split it out into a separate function.

fhahn commented on compiler-rt/lib/tysan/tysan.cpp:
```diff
@@ -153,8 +142,51 @@ static bool isAliasingLegalUp(tysan_type_descriptor *TDA,
+static bool isAliasingLegalWithOffset(tysan_type_descriptor *TDA,
+                                      tysan_type_descriptor *TDB,
+                                      int OffsetB) {
+  // This is handled in the other cases
```
slightly more descriptive?

```suggestion
  // This is handled by calls to isAliasingLegalUp.`
```

fhahn commented on compiler-rt/lib/tysan/tysan.cpp:
```diff
@@ -153,8 +142,51 @@ static bool isAliasingLegalUp(tysan_type_descriptor *TDA,
+  if (TDB->Tag == TYSAN_MEMBER_TD)
+    return false;
+
+  int OffsetA = 0;
```
can this be 

```suggestion
  uptr OffsetA = 0;
```

to be consistent with `isAliasingLegalUp`?

fhahn commented on compiler-rt/lib/tysan/tysan.cpp:
```diff
@@ -153,8 +142,51 @@ static bool isAliasingLegalUp(tysan_type_descriptor *TDA,
 static bool isAliasingLegal(tysan_type_descriptor *TDA,
-                            tysan_type_descriptor *TDB, int TDAOffset = 0) {
+                            tysan_type_descriptor *TDB,
+                            int OffsetB = 0) {
```
can this be


```suggestion
                            uptr OffsetB = 0) {
```

fhahn commented on compiler-rt/lib/tysan/tysan.cpp:
```diff
@@ -102,20 +102,10 @@ static tysan_type_descriptor *getRootTD(tysan_type_descriptor *TD) {
-    TDA = TDA->Member.Base;
-  }
-
+// Walk up TDA to see if it reaches TDB
```
```suggestion
// Walk up TDA to see if it reaches TDB.
```

fhahn commented on compiler-rt/lib/tysan/tysan.cpp:
```diff
@@ -153,8 +143,50 @@ static bool isAliasingLegalUp(tysan_type_descriptor *TDA,
   return false;
 }
 
+static bool isAliasingLegalUp(tysan_type_descriptor *TDA,
+                              tysan_type_descriptor *TDB) {
+  // Walk up the tree starting with TDA to see if we reach TDB.
```
Should the comment be moved outside function?
```suggestion
// Walk up the tree starting with TDA to see if we reach TDB.
static bool isAliasingLegalUp(tysan_type_descriptor *TDA,
                              tysan_type_descriptor *TDB) {
```

fhahn reviewed (approved):
LGTM, thanks

fhahn commented on compiler-rt/test/tysan/inherited_member.cpp:
```diff
@@ -0,0 +1,21 @@
+#include <stdio.h>
+
+class Base {
+public:
+  void *first;
+  void *second;
+  void *third;
```
Oh one thing I forgot was if we can add a test with a type violation that triggers with the new path? Or is that already covered with existing tests?

gbMattN replied:
Its covered in basic.c when we partially access a short

gbMattN committed:
[TySan] Fix false positives with derrived classes
